library/ssl_tls13_server.c

```diff
@@ -2579,18 +2710,267 @@ static int ssl_tls13_write_server_finished( mbedtls_ssl_context *ssl )
         return( ret );
     }
 
-    MBEDTLS_SSL_DEBUG_MSG( 1, ( "Switch to handshake keys for inbound traffic" ) );
-    mbedtls_ssl_set_inbound_transform( ssl, ssl->handshake->transform_handshake );
+    MBEDTLS_SSL_DEBUG_MSG(
+        1, ( "Switch to application keys for outbound traffic"
+                 "( K_send = application )" ) );
+    mbedtls_ssl_set_outbound_transform( ssl, ssl->transform_application );
+
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+    if( mbedtls_ssl_tls13_early_data_is_accepted( ssl ) )
+    {
+        /* TODO: compute early transform here? */
+        MBEDTLS_SSL_DEBUG_MSG(
+            1, ( "Switch to early keys for inbound traffic. "
+                     "( K_recv = early data )" ) );
+        mbedtls_ssl_set_inbound_transform(
+            ssl, ssl->handshake->transform_earlydata );
+        mbedtls_ssl_handshake_set_state( ssl, MBEDTLS_SSL_END_OF_EARLY_DATA );
+        return( 0 );
+    }
+#endif /* MBEDTLS_SSL_EARLY_DATA */
+
+    MBEDTLS_SSL_DEBUG_MSG(
+        1, ( "Switch to handshake keys for inbound traffic"
+                    "( K_recv = handshake )" ) );
+    mbedtls_ssl_set_inbound_transform(
+        ssl, ssl->handshake->transform_handshake );
+    mbedtls_ssl_handshake_set_state( ssl, MBEDTLS_SSL_WAIT_FLIGHT2 );
+
+    return( 0 );
+}
+
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+
+int mbedtls_ssl_read_early_data( mbedtls_ssl_context *ssl,
```

Comment:
continue discussion https://github.com/Mbed-TLS/mbedtls/pull/6537#discussion_r1034505159

Reply:
> I was even thinking at something like the following (not completely sure it can work like that though) where the reading of the early data is done at the level of `mbedtls_ssl_read_early_data()`, not inside `mbedtls_ssl_handshake()`.
> 
> ```c
> int mbedtls_ssl_read_early_data( ssl, buf, size )                               
> {                                                                               
>     do
>     {    
>         /* Case of successive calls to `mbedtls_ssl_read_early_data()` to read early data */
>         if( "early data available" )                                                 
>         {                                                                           
>             ret = ssl_read_real( ssl, buf, size );                                  
>             if( ret < 0 )
>                 return( ret );     
>             return( ret );                                                 
>         }                                                                       
>         
>         ret = mbedtls_ssl_handshake();                                
>         if( ! "can read early data" )
>         {                                           
>             if( ret < 0 )
>                 return( ret );
>         }
>     } while(  "can read early data" );
>     
>     return ”can not read";                                                      
> }                                                                               
> ```
> 

I like your version @ronald-cron-arm, it looks better. I will try it and clear uncertain issues for me.

Bellow are uncertain issues for me, they todo items for myself. I don't expect an answer, but it's better to have an answer.
- [ ] `ret = mbedtls_ssl_handshake` should be `ret = mbedtls_ssl_handshake_step() `
- [ ] `ret = ssl_read_real( ssl, buf, size ); ` , at this point, EOED message or APP message will be received. How to process them? Can `ssl_read_real` handle those cases?
- [ ] `"early data available"`, how to detect it.